File: IsometricRPG/IsometricRPG/World/StructureGenerator.swift
import CoreGraphics

/// Generates procedural structures for chunks
enum StructureGenerator {

    /// Generate structures for a chunk based on biome and seed
    static func generateStructures(
        seed: UInt64,
        biome: Biome,
        elevation: [[Int]],
        existingRooms: [Room]
    ) -> [Structure] {
        var rng = SeededRNG(seed: seed &+ 0x5TRUCTUR3)
        var structures: [Structure] = []

        // Filter templates by biome
        let validTemplates = StructureTemplates.all.filter {
            $0.biomes.contains(biome)
        }

        // Calculate structure budget for this chunk
        let budget = structureBudget(for: biome, rng: &rng)

        // Weighted random selection and placement
        var remainingBudget = budget
        while remainingBudget > 0 && rng.nextFloat() < 0.6 {
            guard let template = weightedRandomTemplate(
                from: validTemplates,
                rng: &rng
            ) else { break }

            // Find valid placement
            if let placement = findPlacement(
                template: template,
                elevation: elevation,
                existing: structures,
                rooms: existingRooms,
                rng: &rng
            ) {
                structures.append(placement)
                remainingBudget -= 1

                // Handle clustering (Phase 5)
                if let clusterSize = template.clusterSize, clusterSize > 1 {
                    structures.append(contentsOf: placeCluster(
                        around: placement,
                        template: template,
                        count: clusterSize - 1,
                        elevation: elevation,
                        existing: structures,
                        rooms: existingRooms,
                        rng: &rng
                    ))
                }
            }
        }

        return structures
    }

    // MARK: - Structure Budget

    private static func structureBudget(for biome: Biome, rng: inout SeededRNG) -> Int {
        switch biome {
        case .forest:
            return 8 + Int(rng.next() % 5)  // 8-12 structures
        case .desert:
            return 3 + Int(rng.next() % 3)  // 3-5 structures
        case .dungeon:
            return 0  // Dungeons use room system, no structures
        case .swamp:
            return 5 + Int(rng.next() % 3)  // 5-7 structures
        case .snow:
            return 4 + Int(rng.next() % 3)  // 4-6 structures
        }
    }

    // MARK: - Template Selection

    private static func weightedRandomTemplate(
        from templates: [StructureTemplate],
        rng: inout SeededRNG
    ) -> StructureTemplate? {
        guard !templates.isEmpty else { return nil }

        let totalWeight = templates.reduce(0) { $0 + $1.spawnWeight }
        var random = rng.nextFloat() * totalWeight

        for template in templates {
            random -= template.spawnWeight
            if random <= 0 {
                return template
            }
        }

        return templates.last
    }

    // MARK: - Placement Validation

    private static func findPlacement(
        template: StructureTemplate,
        elevation: [[Int]],
        existing: [Structure],
        rooms: [Room],
        rng: inout SeededRNG
    ) -> Structure? {
        let maxAttempts = 30
        let size = Constants.chunkSize

        for _ in 0..<maxAttempts {
            let x = Int(rng.next() % UInt64(max(size - template.size.width, 1)))
            let y = Int(rng.next() % UInt64(max(size - template.size.height, 1)))

            // Check 1: Doesn't overlap existing structures
            let candidate = Structure(
                template: template,
                x: x, y: y,
                rotation: 0,
                seed: rng.next()
            )
            if existing.contains(where: { candidate.overlaps($0) }) {
                continue
            }

            // Check 2: Doesn't overlap dungeon rooms
            if rooms.contains(where: { candidate.overlaps($0) }) {
                continue
            }

            // Check 3: Elevation constraints
            if !meetsElevationRequirements(
                template: template,
                x: x, y: y,
                elevation: elevation
            ) {
                continue
            }

            // Check 4: Flatness requirement
            if template.requiresFlat && !isFlat(
                x: x, y: y,
                width: template.size.width,
                height: template.size.height,
                elevation: elevation
            ) {
                continue
            }

            // Valid placement found
            let rotation = Int(rng.next() % 4) * 90
            return Structure(
                template: template,
                x: x, y: y,
                rotation: rotation,
                seed: rng.next()
            )
        }

        return nil
    }

    private static func meetsElevationRequirements(
        template: StructureTemplate,
        x: Int, y: Int,
        elevation: [[Int]]
    ) -> Bool {
        // Get average elevation of structure footprint
        var totalElev = 0
        var count = 0
        for row in y..<min(y + template.size.height, elevation.count) {
            for col in x..<min(x + template.size.width, elevation[0].count) {
                totalElev += elevation[row][col]
                count += 1
            }
        }
        guard count > 0 else { return false }
        let avgElev = totalElev / count

        // Check min elevation
        if let minElev = template.minElevation, avgElev < minElev {
            return false
        }

        // Check max elevation
        if let maxElev = template.maxElevation, avgElev > maxElev {
            return false
        }

        return true
    }

    private static func isFlat(
        x: Int, y: Int,
        width: Int, height: Int,
        elevation: [[Int]]
    ) -> Bool {
        guard y >= 0 && x >= 0 &&
              y < elevation.count && x < elevation[0].count else {
            return false
        }

        let baseElev = elevation[y][x]
        for row in y..<min(y + height, elevation.count) {
            for col in x..<min(x + width, elevation[0].count) {
                if abs(elevation[row][col] - baseElev) > Constants.flatnessThreshold {
                    return false
                }
            }
        }
        return true
    }

    // MARK: - Clustering (Phase 5)

    private static func placeCluster(
        around anchor: Structure,
        template: StructureTemplate,
        count: Int,
        elevation: [[Int]],
        existing: [Structure],
        rooms: [Room],
        rng: inout SeededRNG
    ) -> [Structure] {
        var cluster: [Structure] = []
        let radius = 8  // Cluster within 8 tiles

        for _ in 0..<count {
            // Find position near anchor
            let angle = CGFloat(rng.nextFloat()) * .pi * 2
            let distance = CGFloat(2 + Int(rng.next() % UInt64(radius - 2)))
            let offsetX = Int(cos(angle) * distance)
            let offsetY = Int(sin(angle) * distance)

            let x = anchor.x + offsetX
            let y = anchor.y + offsetY

            // Validate bounds
            guard x >= 0 && y >= 0 &&
                  x + template.size.width <= Constants.chunkSize &&
                  y + template.size.height <= Constants.chunkSize else {
                continue
            }

            // Check placement validity
            let candidate = Structure(
                template: template,
                x: x, y: y,
                rotation: Int(rng.next() % 4) * 90,
                seed: rng.next()
            )

            // Check overlaps
            if existing.contains(where: { candidate.overlaps($0) }) {
                continue
            }
            if cluster.contains(where: { candidate.overlaps($0) }) {
                continue
            }
            if rooms.contains(where: { candidate.overlaps($0) }) {
                continue
            }

            // Check elevation requirements
            if !meetsElevationRequirements(
                template: template,
                x: x, y: y,
                elevation: elevation
            ) {
                continue
            }

            // Check flatness
            if template.requiresFlat && !isFlat(
                x: x, y: y,
                width: template.size.width,
                height: template.size.height,
                elevation: elevation
            ) {
                continue
            }

            cluster.append(candidate)
        }

        return cluster
    }

    // MARK: - Advanced Placement (Phase 5)

    static func findHighGround(
        elevation: [[Int]],
        rng: inout SeededRNG
    ) -> (x: Int, y: Int)? {
        // Find positions with elevation > 100
        var highPositions: [(Int, Int)] = []
        for row in 0..<elevation.count {
            for col in 0..<elevation[0].count {
                if elevation[row][col] > 100 {
                    highPositions.append((col, row))
                }
            }
        }

        guard !highPositions.isEmpty else { return nil }
        let index = Int(rng.next() % UInt64(highPositions.count))
        return highPositions[index]
    }
}

File: IsometricRPGMono/World/StructureGenerator.cs
namespace IsometricRPG;

public static class StructureGenerator
{
    /// <summary>
    /// Generate structures for a chunk given its world-tile origin and an elevation map.
    /// </summary>
    /// <param name="chunkOriginX">World X tile of chunk origin.</param>
    /// <param name="chunkOriginY">World Y tile of chunk origin.</param>
    /// <param name="elevation">Elevation values indexed [row, col] within the chunk.</param>
    /// <param name="biome">Dominant biome of the chunk.</param>
    /// <param name="rng">Seeded RNG shared with the world generator.</param>
    public static List<Structure> GenerateForChunk(
        int chunkOriginX,
        int chunkOriginY,
        float[,] elevation,
        Biome biome,
        SeededRNG rng)
    {
        var structures = new List<Structure>();

        // Pick candidate templates allowed in this biome
        var candidates = StructureTemplates.All
            .Where(t => t.AllowedBiomes.Contains(biome))
            .ToList();

        if (candidates.Count == 0)
            return structures;

        // Attempt to place a small number of structures per chunk
        int maxAttempts = 5;
        for (int attempt = 0; attempt < maxAttempts; attempt++)
        {
            // Weighted random template selection
            var template = SelectWeightedTemplate(candidates, rng);
            if (template == null) break;

            // Find a suitable placement position
            var placement = FindHighGround(
                elevation,
                template.Size.Width,
                template.Size.Height,
                template.MinElevation,
                template.MaxElevation,
                rng);

            if (placement == null) continue;

            var (localX, localY) = placement.Value;
            int worldX = chunkOriginX + localX;
            int worldY = chunkOriginY + localY;

            // Avoid stacking on top of existing structures
            bool overlaps = structures.Any(s =>
                worldX < s.WorldX + s.Size.Width  &&
                worldX + template.Size.Width  > s.WorldX &&
                worldY < s.WorldY + s.Size.Height &&
                worldY + template.Size.Height > s.WorldY);

            if (overlaps) continue;

            structures.Add(new Structure(template, worldX, worldY));
        }

        return structures;
    }

    /// <summary>
    /// Generate a cluster of tree/natural structures around a focal point.
    /// Uses trigonometry to spread placements in a rough circle.
    /// </summary>
    public static List<Structure> GenerateCluster(
        int centerWorldX,
        int centerWorldY,
        float radius,
        Biome biome,
        SeededRNG rng,
        int count = 4)
    {
        var structures = new List<Structure>();

        var candidates = StructureTemplates.All
            .Where(t => t.AllowedBiomes.Contains(biome) && t.Category == StructureCategory.Natural)
            .ToList();

        if (candidates.Count == 0)
            return structures;

        for (int i = 0; i < count; i++)
        {
            var template = SelectWeightedTemplate(candidates, rng);
            if (template == null) continue;

            // Spread placements around the circle
            float angle = (float)rng.NextFloat() * MathF.PI * 2f;
            float dist  = (float)rng.NextFloat() * radius;

            int offsetX = (int)(MathF.Cos(angle) * dist);
            int offsetY = (int)(MathF.Sin(angle) * dist);

            int worldX = centerWorldX + offsetX;
            int worldY = centerWorldY + offsetY;

            bool overlaps = structures.Any(s =>
                worldX < s.WorldX + s.Size.Width  &&
                worldX + template.Size.Width  > s.WorldX &&
                worldY < s.WorldY + s.Size.Height &&
                worldY + template.Size.Height > s.WorldY);

            if (!overlaps)
                structures.Add(new Structure(template, worldX, worldY));
        }

        return structures;
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    private static StructureTemplate? SelectWeightedTemplate(
        List<StructureTemplate> candidates, SeededRNG rng)
    {
        if (candidates.Count == 0) return null;

        float totalWeight = candidates.Sum(t => t.SpawnWeight);
        float roll = (float)rng.NextFloat() * totalWeight;

        float cumulative = 0f;
        foreach (var t in candidates)
        {
            cumulative += t.SpawnWeight;
            if (roll <= cumulative)
                return t;
        }

        return candidates[candidates.Count - 1];
    }

    /// <summary>
    /// Searches the elevation grid for a flat region large enough for the template.
    /// Returns (localX, localY) — top-left corner within the chunk — or null.
    /// </summary>
    private static (int x, int y)? FindHighGround(
        float[,] elevation,
        int width,
        int height,
        int minElev,
        int maxElev,
        SeededRNG rng)
    {
        int rows = elevation.GetLength(0);
        int cols = elevation.GetLength(1);

        // Scale thresholds to 0-1 float range
        float minF = minElev / 100f;
        float maxF = maxElev / 100f;

        // Gather valid flat regions
        var validPositions = new List<(int x, int y)>();

        for (int row = 0; row <= rows - height; row++)
        {
            for (int col = 0; col <= cols - width; col++)
            {
                bool valid = true;
                float baseElev = elevation[row, col];

                if (baseElev < minF || baseElev > maxF)
                    continue;

                for (int dr = 0; dr < height && valid; dr++)
                    for (int dc = 0; dc < width && valid; dc++)
                    {
                        float e = elevation[row + dr, col + dc];
                        if (e < minF || e > maxF) { valid = false; break; }
                        if (MathF.Abs(e - baseElev) > 0.05f) { valid = false; break; }
                    }

                if (valid)
                    validPositions.Add((col, row));
            }
        }

        if (validPositions.Count == 0) return null;

        int idx = (int)(rng.Next() % (ulong)validPositions.Count);
        return validPositions[idx];
    }
}
